### PR TITLE
Replace some `mem::forget`'s with `ManuallyDrop`

### PR DESCRIPTION
**Pull request by GrigorenkoPV**
> but I would like to see a larger effort to replace all uses of `mem::forget`.

_Originally posted by @saethlin in https://github.com/rust-lang/rust/issues/127584#issuecomment-2226087767_

So,
r? @saethlin

Sorry, I have finished writing all of this before I got your response.

### CODE REVIEW

**Comment by workingjubilee** [comment]
@GrigorenkoPV Hello, can you split this into PRs for the compiler and library?

**Comment by GrigorenkoPV** [comment]
> @GrigorenkoPV Hello, can you split this into PRs for the compiler and library?

Sure. I think I will just drop the compiler changes entirely since there were only two minuscule ones and the initial discussion was about libs only anyways.

**Comment by workingjubilee** [comment]
sounds good to me!

**Comment by workingjubilee** [comment]
r? libs

File: library/std/src/sys/pal/wasi/thread.rs [hunk 1]
```diff
@@ -173,8 +173,7 @@ impl Thread {
         cfg_if::cfg_if! {
             if #[cfg(target_feature = "atomics")] {
                 unsafe {
-                    let ret = libc::pthread_join(self.id, ptr::null_mut());
-                    mem::forget(self);
+                    let ret = libc::pthread_join(mem::ManuallyDrop::new(self).id, ptr::null_mut());
```
Comment (Amanieu) [inline library/std/src/sys/pal/wasi/thread.rs:176]: Split this into 2 lines with a `let id = ...`.

File: library/std/src/sys/pal/sgx/abi/usercalls/alloc.rs [hunk 2]
```diff
@@ -266,9 +267,8 @@ where
     /// Converts this value into a raw pointer. The value will no longer be
     /// automatically freed.
     pub fn into_raw(self) -> *mut T {
-        let ret = self.0;
-        mem::forget(self);
-        ret.as_ptr() as _
+        let Self(nonnull) = self;
```
Comment (Amanieu) [inline library/std/src/sys/pal/sgx/abi/usercalls/alloc.rs:270]: The inner `NonNull` is `Copy` so I think `self` still gets dropped after this line.

File: library/std/src/sys/pal/wasi/thread.rs (hunk 1 above)
Comment (GrigorenkoPV) [reply library/std/src/sys/pal/wasi/thread.rs:176]: Sure! I have done this for other platforms, but for some reason missed this one.

File: library/std/src/sys/pal/sgx/abi/usercalls/alloc.rs (hunk 2 above)
Comment (GrigorenkoPV) [reply library/std/src/sys/pal/sgx/abi/usercalls/alloc.rs:270]: Oh, you are right! I have totally forgotten how destructuring works.

**Pull request merged**